### PR TITLE
Add external data format support

### PR DESCRIPTION
**Pull request by kahmed10**

### CODE REVIEW

File: src/include/migraphx/read_buffer.hpp [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+#include <migraphx/config.hpp>
+#include <migraphx/errors.hpp>
+#include <fstream>
+#include <vector>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+std::vector<char> read_buffer(const std::string& filename);
```
Comment (pfultz2) [inline src/include/migraphx/read_buffer.hpp:12]: This should go into a file_buffer.cpp file so its consistent with PR#664.

File: src/onnx/onnx.cpp [hunk 2]
```diff
@@ -2568,8 +2571,18 @@ struct onnx_parser
...
     {
+        path = ".";
+        if(contains(filename, '/'))
+            path = filename.substr(0, filename.find_last_of('/'));
+    }
+
+    void parse_from(std::istream& is, std::string name = "")
+    {
+        this->filename = std::move(name);
+        get_abs_path();
```
Comment (pfultz2) [inline src/onnx/onnx.cpp:2584]: We should probably use [`std::filesystem`](https://en.cppreference.com/w/cpp/filesystem) for this. In PR#664, I added a migraphx/filesystem.hpp header to select the filesystem available. Everything is put in the `migraphx::fs` namespace. So you should be able to compute the path with `fs::path(this->filename).parent_path()`.

File: src/onnx/onnx.cpp [hunk 3]
```diff
@@ -52,6 +53,8 @@ namespace onnx = onnx_for_migraphx;
 
 struct onnx_parser
 {
+    std::string filename;
+    std::string path;
```
Comment (scxiao) [inline src/onnx/onnx.cpp:57]: is there a default value for the filename and path like the onnx file name appending something?

File: src/file_buffer.cpp [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+#include <fstream>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+std::vector<char> read_buffer(const std::string& filename)
+{
+    std::ifstream is(filename, std::ios::binary | std::ios::ate);
+    if(not is.good())
+        MIGRAPHX_THROW("Bad stream: " + filename);
```
[the remark this answers is not in the document]
Comment (kahmed10) [reply src/file_buffer.cpp:12]: The function name is now embedded into `MIGRAPHX_THROW`, so I think it's redundant

File: src/onnx/onnx.cpp (hunk 3 above)
Comment (kahmed10) [reply src/onnx/onnx.cpp:57]: I assign it when I call `get_abs_path`, but I can change it to be "." by default. The assignment will change once PR #664 is merged in.

File: src/file_buffer.cpp [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    if(!is.read(buffer.data(), size))
+    {
+        MIGRAPHX_THROW("Error reading file: " + filename);
+    }
+    return buffer;
+}
+
+void write_buffer(const std::string& filename, const char* buffer, std::size_t size)
+{
+    std::ofstream os(filename);
```
[the remark this answers is not in the document]
Comment (kahmed10) [reply src/file_buffer.cpp:28]: deepcode did not give an error for this one, so it might not be necessary

**Review by scxiao — APPROVED** [review]

**Comment by scxiao** [comment]
@kahmed10 You need to resolve the merge conflict for this one

**Review by pfultz2 — APPROVED** [review]

**Review by turneram — APPROVED** [review]

**Comment by mvermeulen** [comment]
Any clues on what is going on in Jenkins for this one... hoping to have this be next merged in.

**Pull request merged**